include/revng/Pipeline/Target.h

```diff
@@ -0,0 +1,359 @@
+                                   "string %s was not in expected form "
+                                   "<path:kind>",
+                                   AsString.str().c_str());
+
+  llvm::SmallVector<llvm::StringRef, 4> Path;
+  Parts[0].split(Path, '/');
+
+  auto It = llvm::find_if(Dict,
+                          [&Parts](Kind &K) { return Parts[1] == K.name(); });
+  if (It == Dict.end())
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "No known Kind %s in dictionary",
+                                   Parts[1].str().c_str());
+
+  return Target(std::move(Path), *It);
+}
+
+/// a sorted list of targets.
+class TargetsList {
+public:
+  using List = llvm::SmallVector<Target, 4>;
+  using iterator = List::iterator;
+  using const_iterator = List::const_iterator;
+  using value_type = List::value_type;
+  using size_type = List::size_type;
+  using reference = List::reference;
+  using pointer = List::pointer;
+
+private:
+  List Contained;
+
+public:
+  TargetsList() = default;
+  TargetsList(List C) : Contained(std::move(C)) { removeDuplicates(); }
+
+  iterator begin() { return Contained.begin(); }
+
+  iterator end() { return Contained.end(); }
+
+  const_iterator begin() const { return Contained.begin(); }
+
+  const_iterator end() const { return Contained.end(); }
+
+  size_t size() const { return Contained.size(); }
+
+  template<typename... Args>
+  void emplace_back(Args &&...args) {
+    Contained.emplace_back(std::forward<Args>(args)...);
+    removeDuplicates();
+  }
+
+  void dump() const debug_function { dump(dbg); }
+
+  template<typename OStream>
+  void dump(OStream &OS, size_t Indents = 0) const {
+    for (const auto &Entry : Contained)
+      Entry.dump(OS, Indents);
+  }
+
+  void merge(const TargetsList &Other);
+
+  void push_back(const Target &Target) {
+    Contained.push_back(Target);
+    removeDuplicates();
+  }
+
+  const Target &operator[](size_t Index) const { return Contained[Index]; }
+
+  Target &operator[](size_t Index) { return Contained[Index]; }
+
+  template<typename... Args>
+  auto erase(Args &&...args) {
+    return Contained.erase(std::forward<Args>(args)...);
+  }
+
+  Target &front() { return Contained.front(); }
+
+  const Target &front() const { return Contained.front(); }
+
+  bool contains(const Target &Target) const;
+
+  bool contains(const TargetsList &Targets) const {
+    return llvm::all_of(Targets, [this](const Target &Target) {
+      return contains(Target);
+    });
+  }
+
+  TargetsList filter(const Kind &K) const {
+    List C;
+    for (const auto &Target : Contained)
+      if (&Target.getKind() == &K)
+        C.push_back(Target);
+
+    return TargetsList(std::move(C));
+  }
+
+  bool empty() const { return Contained.empty(); }
+
+private:
+  void removeDuplicates();
+};
+
+/// a map from container name to target list that
+/// is usually used to represents the state of a step.
+class ContainersStatus {
```

Comment:
This data structure is suspicious. It's 80 lines but:
* it doesn't seem to bring much benefit w.r.t. `using ContainerStatus = llvm::StringMap<TargetsList>`.
* `operator[]` has an unexpected behavior
* `at() const` does not assert in case of absence.
* `size() != std::distance(begin(), end()`

The `dump` method is useful but could become a free function.

I suggest:

* Switch to `using ContainerStatus = llvm::StringMap<TargetsList>` + free `dumpContainerStatus` function
* `class ContainersStatus : llvm::StringMap<TargetsList> {  void dump(); };

Reply:
eh, this too has  a merge method, i can have them as free function, but i would prefer hide the implementations detail of this class.

Reply:
I suggest you go for inheritance from `llvm::StringMap<TargetsList>` and add the methods you'd like.
If you feel you're exposing too much inheriting `StringMap`, see if you can make an intermediate class that exposes what you want, but I'd really like this class to be simple and only add methods that are not those of a standard map.
As it is suspiciously suspicious.

Reply:
the issue here is similar to container set one, the point is that operator[] is weird because this data structure is lazy, and should generate the value both when const operartor[] and the non const one are used. 

The correct fix would be to use mutable, and to hide entirely the detectable lazyness of this data structure. this would allow to  have proper at and operator[] methods.

Reply:
it can't be a temporary, it would be taking the address of a stack object. or i understand wrong what you meant?

Reply:
I'm dumb.

What I wanted to say: let's have a `static MyClass Empty;` field and return that.

Reply:
i thought about this, the issue with it is that the non const overload cannot return a ref to this object, and then you would have that the const and non const version return different objects, which is really really strange

Reply:
Yeah, we'd have `&static_cast<const MyClass &>(Obj).get(X) != &Obj.get(X)`.
But do we care?
Feels more elegant than `mutable`...

The alternative is to drop all the `const` in users and that's it.